pinecone/client.go

```diff
@@ -20,3 +41,357 @@ func (c *Client) IndexWithNamespace(host string, namespace string) (*IndexConnec
 	}
 	return idx, nil
 }
+
+func (c *Client) ListIndexes(ctx context.Context) ([]*Index, error) {
+	res, err := c.restClient.ListIndexes(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", res.StatusCode)
+	}
+
+	var indexList control.IndexList
+	err = json.NewDecoder(res.Body).Decode(&indexList)
+	if err != nil {
+		return nil, err
+	}
+
+	indexes := make([]*Index, len(*indexList.Indexes))
+	for i, idx := range *indexList.Indexes {
+		indexes[i] = toIndex(&idx)
+	}
+
+	return indexes, nil
+}
+
+type CreatePodIndexRequest struct {
+	Name             string
+	Dimension        int32
+	Metric           IndexMetric
+	Environment      string
+	PodType          string
+	Shards           int32
+	Replicas         int32
+	SourceCollection *string
+	MetadataConfig   *PodSpecMetadataConfig
+}
+
+func (req CreatePodIndexRequest) ReplicaCount() *int32 {
+	x := minOne(req.Replicas)
+	return &x
+}
+
+func (req CreatePodIndexRequest) ShardCount() *int32 {
+	x := minOne(req.Shards)
+	return &x
+}
+
+func (req CreatePodIndexRequest) TotalCount() *int {
+	x := int(*req.ReplicaCount() * *req.ShardCount())
+	return &x
+}
+
+func (c *Client) CreatePodIndex(ctx context.Context, in *CreatePodIndexRequest) (*Index, error) {
+	metric := control.IndexMetric(in.Metric)
+	req := control.CreateIndexRequest{
+		Name:      in.Name,
+		Dimension: in.Dimension,
+		Metric:    &metric,
+	}
+
+	//add the spec to req.
+	//because this is defined as an anon struct in the generated code, it must match exactly here.
+	req.Spec = control.CreateIndexRequest_Spec{
+		Pod: &struct {
+			Environment    string `json:"environment"`
+			MetadataConfig *struct {
+				Indexed *[]string `json:"indexed,omitempty"`
+			} `json:"metadata_config,omitempty"`
+			PodType          control.PodSpecPodType   `json:"pod_type"`
+			Pods             *int                     `json:"pods,omitempty"`
+			Replicas         *control.PodSpecReplicas `json:"replicas,omitempty"`
+			Shards           *control.PodSpecShards   `json:"shards,omitempty"`
+			SourceCollection *string                  `json:"source_collection,omitempty"`
+		}{
+			Environment:      in.Environment,
+			PodType:          in.PodType,
+			Pods:             in.TotalCount(),
+			Replicas:         in.ReplicaCount(),
+			Shards:           in.ShardCount(),
```

Comment:
My understanding of this and the methods above is we're basically massaging the values passed and setting them to 1 if negative. I had initially thought it was for argument validation.

Are we relying on the server for validations in general?

Reply:
Another unfortunate case. These are optional params. If you don't set them in the `in` struct, they are the "zero value" for that type. In most cases, the "zero value" for the type works out OK (empty string, 0, false, etc). However, the API defaults these to 1. So, if the caller of the SDK omits it (or sets it to the zero value) we need to ensure the default is passed in.

The other annoyance that is cleaned up here is one of general "Pod" math by our API.  Pods must equal "Replicas * Shards" where Replicas and Shards are both ints with a min value of 1. This only accepts Replicas and Shards as inputs, omitting the ability to miscalculate Pods.

Reply:
Thanks for the explanation, I actually really like the enforcement of "Pod" math through this.